cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java

```diff
@@ -109,23 +113,78 @@ public static String extractTermValue(String term) {
   }
 
   private static QueryTerm parseQueryTerm(String term) {
-    SearchType parsedSearchType = getSearchType(term);
+    Qualifier parsedQualifier = hasRequirementPrefix(term) ? Qualifier.REQUIRED : Qualifier.OPTIONAL;
+    String parsedTerm = hasRequirementPrefix(term)
+      ? term.substring(1) : term;
+    if (parsedTerm.startsWith("DATE:")) {
+      return parseDateTerm(parsedTerm, parsedQualifier);
+    }
 
-    Qualifier parsedQualifier = hasRequirementPrefix(term) || parsedSearchType == SearchType.NUMERIC
-        ? Qualifier.REQUIRED : Qualifier.OPTIONAL;
+    SearchType parsedSearchType = getSearchType(parsedTerm);
 
-    String parsedTerm = hasRequirementPrefix(term)
-        ? term.substring(1) : term;
+    parsedQualifier = parsedSearchType == SearchType.NUMERIC || parsedSearchType == SearchType.DATE
+      ? Qualifier.REQUIRED : parsedQualifier;
 
     Comparison parsedComparison = parsedSearchType == SearchType.STRING
         ? Comparison.EQUALS : getComparison(term);
 
+    if (parsedSearchType.equals(SearchType.DATE)) {
+      return new QueryTerm(parsedTerm, parsedQualifier, parsedSearchType, parsedComparison, parseDate(parsedTerm));
+    }
+
     return new QueryTerm(parsedTerm, parsedQualifier, parsedSearchType, parsedComparison);
   }
 
+  /**
+   * Parses a user's query when the "DATE" keyword is detected in the beginning of the query.
+   * Extracts the comparison operator and converts the date string into a Unix timestamp and
+   * creates a QueryTerm with this information.
+   * If the date term cannot be converted to a Unix timestamp, creates a regular QueryTerm with the original term input.
+   *
+   * @param term      an individual term from the user's original query
+   * @param qualifier the qualifier that is detected by queryParser
+   * @return a date QueryTerm with the extracted information. If date query is invalid, returns a term-search QueryTerm
+   */
+  private static QueryTerm parseDateTerm(String term, Qualifier qualifier) {
+    String valueTerm = lastSubTerm(term);
+    String dateTerm = extractTermValue(valueTerm);
+    Long date = parseDate(dateTerm);
+    // If the user's date cannot be parsed then create a regular QueryTerm to search for the query as a string.
+    // In this case assume that the DATE: keyword is a part of the user's intended string search.
+    if (date == null) {
+      return new QueryTerm(term, qualifier);
+    }
+    // Remove "DATE:" from the term.
+    term = term.substring(5);
+
+    if (term.contains(KEYVALUE_SEPARATOR)) {
+      String[] split = term.split(KEYVALUE_SEPARATOR);
+      if (split.length > 2) {
+        //remove DATE: so that it is not considered as a field name.
```

Comment:
why are we removing `DATE:` here and in line 158?

Reply:
line 164 is actually obsolete now because of line 158, so I can remove that if-statement.
I remove 'DATE:' from the query term so that it is not treated as a field when parsed later on. 

Originally I had thought that if a query was of the form `DATE:field:value` then "DATE:" should be removed because a field is defined, but if the form is `DATE:value` then it is possible that the user meant to have "DATE:" as a field and not a keyword. But eventually I decided to not assume that, and instead assume that 'DATE:' always refers to a date search. But let me know your opinions on that!